eclair-core/src/main/scala/fr/acinq/eclair/channel/TxPublisher.scala

```diff
@@ -0,0 +1,418 @@
+          }
+          run(cltvDelayedTxs, csvDelayedTxs + (p.tx.txid -> TxWithRelativeDelay(p, csvTimeouts.keySet)))
+        } else if (cltvTimeout > blockCount) {
+          log.info(s"delaying publication of txid=${p.tx.txid} until block=$cltvTimeout (current block=$blockCount)")
+          val cltvDelayedTxs1 = cltvDelayedTxs + (cltvTimeout -> (cltvDelayedTxs.getOrElse(cltvTimeout, Seq.empty) :+ p))
+          run(cltvDelayedTxs1, csvDelayedTxs)
+        } else {
+          publish(p)
+          Behaviors.same
+        }
+
+      case ParentTxConfirmed(p, parentTxId) =>
+        log.info(s"parent tx of txid=${p.tx.txid} has been confirmed (parent txid=$parentTxId)")
+        val blockCount = nodeParams.currentBlockHeight
+        csvDelayedTxs.get(p.tx.txid) match {
+          case Some(TxWithRelativeDelay(_, parentTxIds)) =>
+            val txWithRelativeDelay1 = TxWithRelativeDelay(p, parentTxIds - parentTxId)
+            if (txWithRelativeDelay1.parentTxIds.isEmpty) {
+              log.info(s"all parent txs of txid=${p.tx.txid} have been confirmed")
+              val csvDelayedTx1 = csvDelayedTxs - p.tx.txid
+              val cltvTimeout = Scripts.cltvTimeout(p.tx)
+              if (cltvTimeout > blockCount) {
+                log.info(s"delaying publication of txid=${p.tx.txid} until block=$cltvTimeout (current block=$blockCount)")
+                val cltvDelayedTxs1 = cltvDelayedTxs + (cltvTimeout -> (cltvDelayedTxs.getOrElse(cltvTimeout, Seq.empty) :+ p))
+                run(cltvDelayedTxs1, csvDelayedTx1)
+              } else {
+                publish(p)
+                run(cltvDelayedTxs, csvDelayedTx1)
+              }
+            } else {
+              log.info(s"some parent txs of txid=${p.tx.txid} are still unconfirmed (parent txids=${txWithRelativeDelay1.parentTxIds.mkString(",")})")
+              run(cltvDelayedTxs, csvDelayedTxs + (p.tx.txid -> txWithRelativeDelay1))
+            }
+          case None =>
+            log.warn(s"txid=${p.tx.txid} not found for parent txid=$parentTxId")
+            Behaviors.same
+        }
+
+      case WrappedCurrentBlockCount(blockCount) =>
+        val toPublish = cltvDelayedTxs.view.filterKeys(_ <= blockCount)
+        toPublish.values.flatten.foreach(tx => publish(tx))
+        run(cltvDelayedTxs -- toPublish.keys, csvDelayedTxs)
+
+      case PublishNextBlock(p) =>
+        val nextBlockCount = nodeParams.currentBlockHeight + 1
+        val cltvDelayedTxs1 = cltvDelayedTxs + (nextBlockCount -> (cltvDelayedTxs.getOrElse(nextBlockCount, Seq.empty) :+ p))
+        run(cltvDelayedTxs1, csvDelayedTxs)
+    }
+
+  implicit val ec: ExecutionContext = ExecutionContext.global
+
+  // NOTE: we use a single thread to publish transactions so that it preserves order.
+  // CHANGING THIS WILL RESULT IN CONCURRENCY ISSUES WHILE PUBLISHING PARENT AND CHILD TXS!
+  val singleThreadExecutionContext = ExecutionContext.fromExecutor(Executors.newSingleThreadExecutor())
```

Comment:
We can't create a thread per channel, this should be defined in the _companion object_. All transaction publication will be serialized, but it's okay, and that's how it is currently.

Reply:
Good catch

Reply:
Done in https://github.com/ACINQ/eclair/pull/1749/commits/d17b3d0ab1d33f504c04ad67a537b46649ee3230

Reply:
I don't think we should make `singleThreadExecutionContext` a class-wide implicit execution context, because it may be picked up by other unrelated asynchronous tasks and will generate deadlocks. Instead I would not make it an implicit, and have the callers reference it directly from the object, so it is very clear what operations are serialized. Note that in the previous implementation it was actually provided explicitly and still is:

```scala
  private def publish(tx: Transaction): Future[ByteVector32] = {
    client.publishTransaction(tx)(singleThreadExecutionContext).recoverWith {
      case t: Throwable =>
        log.error("cannot publish tx: reason={} txid={}", t.getMessage, tx.txid)
        Future.failed(t)
    }
  }
```

Reply:
It feels a bit dangerous because we can easily forget to use it on one of the calls that require it (and only code review catches that).
Maybe if we force *all* calls to explicitly specify their execution context (no implicit global execution context as we had in previous `ZmqWatcher`) then it's better, I'll try that.

Reply:
> we can easily forget to use it on one of the calls that require it

If there is no implicit execution context it won't compile. However I share the concern, maybe we could go with another architecture that prevents foot guns.

Reply:
I simplified this in https://github.com/ACINQ/eclair/pull/1749/commits/989baaa254dc370d2369619a2efa00de534f9bb7
There's only one place where we call `client.publishTransaction` now and it's where we explicitly use the `singleThreadExecutionContext`.

Unless I'm missing something, I believe the only case where tx ordering matters is for the commit tx, that needs to be published first. There is not parent-child dependency elsewhere since we now publish htlc 3rd-stage txs only once htlc-txs are confirmed. Maybe we can do something more clever in the future where we special case the commit tx somehow and wait for it to be in the mempool before publishing other txs, TBD.

Reply:
We could also use the `pipeTo` pattern and maintain a publication queue in the actor state, so we don't rely on a single thread, which is kind of hacky.

> Unless I'm missing something, I believe the only case where tx ordering matters is for the commit tx, that needs to be published first.

I'm embarrassed to say I can't tell where the ordering is required. Our 2nd stage transactions are delayed relative to the commit tx, so there shouldn't be a problem either. 🤔

Reply:
> We could also use the pipeTo pattern and maintain a publication queue in the actor state, so we don't rely on a single thread, which is kind of hacky.

I'll keep that in mind in a future PR where I'll make the tx publishing logic more modular (in particular to cleanly unlock utxos in case of failures), we may have an opportunity to improve that then.

> I'm embarrassed to say I can't tell where the ordering is required. Our 2nd stage transactions are delayed relative to the commit tx

The only cases where I think it could matter are:

- spending the anchor (which obviously needs the commit tx to be in the mempool)
- HTLC txs with the default commitment format (they don't have a csv 1, they can be in the same block as the commit tx)

Reply:
Doh! Htlcs txs are _not_ delayed relative to the commitment tx. Their _outputs_ are! Updated my comment.